sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala

```diff
@@ -379,364 +401,248 @@ class GpuDynamicPartitionDataSingleWriter(
+      i => Some(bucketIdCol.getInt(i))
+    } else {
+      _ => None
     }
   }
 
-  override def write(batch: ColumnarBatch): Unit = {
-    // this single writer always passes `cachesMap` as None
-    write(batch, cachesMap = None)
-  }
-
-  private case class SplitAndPath(var split: SpillableColumnarBatch, path: String)
-      extends AutoCloseable {
-    override def close(): Unit = {
-      split.safeClose()
-      split = null
+  protected def genGetPartitionPathFunc(keyHostCb: ColumnarBatch): Int => Option[String] = {
+    if (isPartitioned) {
+      // Use the existing code to convert each row into a path. It would be nice to do this
+      // on the GPU, but the data should be small and there are things we cannot easily
+      // support on the GPU right now
+      import scala.collection.JavaConverters._
+      val partCols = description.partitionColumns.indices.map(keyHostCb.column)
+      val iter = new ColumnarBatch(partCols.toArray, keyHostCb.numRows()).rowIterator()
+        .asScala.map(getPartitionPath)
+      _ => Some(iter.next)
+    } else {
+      _ => None
     }
   }
 
   /**
-   * Split a batch according to the sorted keys (partitions). Returns a tuple with an
-   * array of the splits as `ContiguousTable`'s, and an array of paths to use to
-   * write each partition.
+   * Split a batch according to the sorted keys (partitions + bucket ids).
+   * Returns a tuple with an array of the splits as `ContiguousTable`'s, an array of
+   * paths and bucket ids to use to write each partition and(or) bucket file.
    */
-  private def splitBatchByKeyAndClose(
-      batch: ColumnarBatch,
-      partDataTypes: Array[DataType]): Array[SplitAndPath] = {
-    val (outputColumnsBatch, partitionColumnsBatch) = withResource(batch) { _ =>
-      closeOnExcept(getOutputColumnsAsBatch(batch)) { outputColumnsBatch =>
-        closeOnExcept(getPartitionColumnsAsBatch(batch)) { partitionColumnsBatch =>
-          (outputColumnsBatch, partitionColumnsBatch)
+  private def splitBatchByKeyAndClose(batch: ColumnarBatch): Array[SplitPack] = {
+    val cbs@Array(keysCb, dataCb) = withResource(batch) { _ =>
+      Array(getKeysBatch(batch), getDataColumnsAsBatch(batch))
+    }
+    val (keyHostCb, splits) = withResource(cbs) { _ =>
+      val keysCbTypes = GpuColumnVector.extractTypes(keysCb)
+      val (distinctKeysTbl, splitIds) = withResource(keysCb) { _ =>
+        withResource(GpuColumnVector.from(keysCb)) { keysTable =>
+          val distinctKeysTbl = distinctAndSort(keysTable)
+          (distinctKeysTbl, splitIndexes(keysTable, distinctKeysTbl))
         }
       }
-    }
-    val (cbKeys, partitionIndexes) = closeOnExcept(outputColumnsBatch) { _ =>
-      val partitionColumnsTbl = withResource(partitionColumnsBatch) { _ =>
-        GpuColumnVector.from(partitionColumnsBatch)
+      cbs(0) = null // avoid double close
+      val keyHostCb = withResource(distinctKeysTbl) { _ =>
+        copyToHostAsBatch(distinctKeysTbl, keysCbTypes)
```

Comment:
I'm not sure how small this actually is.  We have potentially a few partition columns and optionally a bucket column.  So the number of rows is proportional to the number of partitions we want to bucket on * the carnality of the partition by keys. But if Spark is storing this data in separate directory/file paths, then we have to follow them and do it the same way, which means it should be small or the processing is going to be really painful no matter what.

Reply:
Yeah, maybe we can print a log to tell users that GPU is not good at doing this when the number (= numBuckets * numPartitions) is too big.